developer_guides/uuid/index.rst

```diff
@@ -0,0 +1,133 @@
+
+   /* Component extended tokens */
+   static const struct sof_topology_token comp_ext_tokens[] = {
+    {SOF_TKN_COMP_UUID,
+        SND_SOC_TPLG_TUPLE_TYPE_UUID, get_token_uuid,
+        offsetof(struct sof_ipc_comp_new_ext, uuid), 0},
+   };
+
+.. code-block:: none
+
+   static int sof_widget_ready(struct snd_soc_component *scomp, int index,
+                               struct snd_soc_dapm_widget *w,
+                               struct snd_soc_tplg_dapm_widget *tw)
+   {
+   ...
+           ret = sof_parse_tokens(scomp, &swidget->comp_ext, comp_ext_tokens,
+                                  ARRAY_SIZE(comp_ext_tokens), tw->priv.array,
+                                  le32_to_cpu(tw->priv.size));
+   ...
+   }
+
+UUID Arrays Stored Section
+**************************
+Only the UUID arrays for runtime created components are stored to the
+.rodata section as static data, it will occupy small memory, e.g.
+19 component types * 16 Bytes/component type = 304 Bytes.
+
+UUID to Component Driver Mapping
+********************************
+The firmware will use UUID byte array to match the component driver, if
+it is provided from the Linux driver side, otherwise, fallback to use
+the traditional component type for backward compatible issue.
+
+.. code-block:: none
+
+   static const struct comp_driver *get_drv(struct sof_ipc_comp *comp)
+   {
+        ...
+        /* validate the extended data */
+        ...
+        /* use UUID to match the driver if UUID is provided */
+        if (comp->ext_data_offset) {
+            /* use component type if old tplg without UUID used */
+            if (sof_is_uuid_nil(comp_ext->uuid))
+                goto comp_type_match;
+
+            /* search driver list with UUID */
+            ...
+            /* matched, return drv */
+            return drv;
+
+            /* not found, failed */
+            return NULL;
+        }
+
+   comp_type_match:
+        /* search driver list for driver type */
+        ...
+        /* return the component type matched driver */
+        return drv;
+   }
+
+ABI Alignment for UUID Support
+******************************
+In general, UUID will be used only all FW/tplg/driver are in the new versions.
+
+.. csv-table:: Firmware/tplg/driver ABI alignment for UUID support
+   :header: "Firmware", "Topology file", "Linux driver", "Comments"
+   :widths: 10, 10, 15, 40
+
+   "old", "old", "old", "supported, no UUID"
+   "old", "old", "new", "supported, UUID 0s, FW ignore it"
+   "old", "new", "old", "supported, no UUID"
+   "old", "new", "new", "supported, UUIDs valid, but FW ignore them, component type will be used."
+   "new", "old", "old", "supported, offset/reserved 0 passed to the FW, component type will be used."
+   "new", "old", "new", "supported, UUID 0s, component type will be used in FW."
+   "new", "new", "old", "supported, offset/reserved 0 passed to the FW, component type will be used."
+   "new", "new", "new", "supported, UUID will be supported."
```

Comment:
can we define old and new here?

Reply:
added now.

Reply:
Hmm, this is still a bit hard to read. How about:

UUID will be used in component creation only when support is available across the stack in FW, topology and driver. If an older FW/tplg/driver without UUID support is used, component type information is used instead.

Following table describes the firmware behaviour with different combinations of FW/tplg/driver. 

.. csv-table:: Firmware/tplg/driver ABI alignment for UUID support
   :header: "Firmware", "Topology file", "Linux driver", "Comments"
   :widths: 10, 10, 15, 40

   "no", "no", "no", "supported, no UUID"
   "no", "no", "yes", "supported, UUID 0s, FW ignore it"
   "no", "yes", "no", "supported, no UUID"
   "no", "yes", "yes", "supported, UUIDs valid, but FW ignore them, component type will be used."
   "yes", "no", "no", "supported, offset/reserved 0 passed to the FW, component type will be used."
   "yes", "no", "yes", "supported, UUID 0s, component type will be used in FW."
   "yes", "yes", "no", "supported, offset/reserved 0 passed to the FW, component type will be used."
   "yes", "yes", "yes", "supported, UUID will be supported."